storage/src/adb/keyless.rs

```diff
@@ -162,167 +319,57 @@ impl<E: Storage + Clock + Metrics, V: Codec, H: CHasher> Keyless<E, V, H> {
+                .await?;
 
         // Trim any locations/mmr elements that do not have corresponding operations in log.
-        if aligned_size != has_offset_size {
+        if aligned_size != valid_size {
             warn!(
                 size = aligned_size,
-                new_size = has_offset_size,
+                new_size = valid_size,
                 "trimming locations & mmr elements ahead of log"
             );
-            locations.rewind(has_offset_size).await?;
+            locations.rewind(valid_size).await?;
             locations.sync().await?;
-            mmr.pop((aligned_size - has_offset_size) as usize).await?;
+            mmr.pop((aligned_size - valid_size) as usize).await?;
         }
         assert_eq!(mmr.leaves(), locations.size().await?);
 
         // The tip of the locations journal & mmr, if they exist, must now correspond to a valid log
         // location represented by `section_offset`. We use this as the starting point to replay the
         // log in order to add back any missing items. If they don't exist, we use the very first
         // log operation as our starting point.
-        let (section, offset) = if let Some((s, o)) = section_offset {
-            (s, o)
-        } else {
-            // Since locations & mmr are empty, find the very first log operation as our starting
-            // point. If there is no such operation, it means the db is empty.
-            let op = {
-                let stream = log.replay(0, 0, REPLAY_BUFFER_SIZE).await?;
-                pin_mut!(stream);
-                stream.next().await
-            };
-            let Some(op) = op else {
-                warn!("no starting log operation found, returning empty db");
-                return Ok(Self {
-                    mmr,
-                    log,
-                    size: 0,
-                    locations,
-                    log_items_per_section,
-                    hasher,
-                    last_commit_loc: None,
-                });
-            };
-            // Add the very first log operation to mmr and locations.
-            let op = op?;
-            mmr.add_batched(&mut hasher, &op.3.encode()).await?;
-            mmr.sync(&mut hasher).await?;
-            locations.append(op.1).await?;
-            locations.sync().await?;
-
-            (op.0, op.1)
-        };
-
-        // Replay any log operations that are missing from the other structures, returning the
-        // operation at the log's tip.
-        let last_log_op = {
-            let stream = log.replay(section, offset, REPLAY_BUFFER_SIZE).await?;
-            pin_mut!(stream);
-            let op = stream.next().await;
-            // Very first operation is known to exist & be valid.
-            let op = op.expect("operation known to exist");
-            let mut op = op?.3;
-            while let Some(result) = stream.next().await {
-                let (section, offset, _, next_op) = result?;
```

Comment:
I "think" there was a subtle bug in those code where using `let` here wouldn't update the read offset (and we'd start the replay below from a place much earlier than the last offset)?